File: migration/src/batch.py
import datetime
from src.db import get_db_connection
from src.api import fetch_stock_data, parse_data
from src.signals import calculate_pricemovement, generate_signal

def run_daily_batch():
    conn = get_db_connection()
    # Use dictionary cursor for column access by name
    cursor = conn.cursor(dictionary=True)
    
    print("Starting daily batch processing...")
    
    # 1. Get all tickers from watch_list
    cursor.execute("SELECT ticker FROM watch_list")
    rows = cursor.fetchall()
    tickers = [row['ticker'] for row in rows]
    
    if not tickers:
        print("No tickers found in watch_list.")
        conn.close()
        return

    today_str = datetime.date.today().strftime('%Y-%m-%d')
    
    for ticker in tickers:
        print(f"Processing {ticker}...")
        
        # 2. Fetch data
        raw_data = fetch_stock_data(ticker)
        if not raw_data:
            print(f"Skipping {ticker} due to fetch error.")
            continue
            
        data = parse_data(raw_data)
        
        # 3. Calculate indicators
        current_price = data.get('currentPrice')
        prev_close_val = data.get('previousCloseVal')
        
        if current_price is None or prev_close_val is None:
             print(f"Skipping {ticker}: Missing price data.")
             continue
             
        movement = calculate_pricemovement(current_price, prev_close_val)
        signal = generate_signal(movement)
        
        # 4. Update watch_list
        # MySQL uses %s placeholder
        cursor.execute('''
            UPDATE watch_list
            SET currentPrice = %s,
                previousClose = %s,
                dividendYield = %s,
                per = %s,
                pbr = %s,
                marketCap = %s,
                volume = %s,
                pricemovement = %s,
                `signal_val` = %s
            WHERE ticker = %s
        ''', (
            current_price,
            data.get('previousClose'),
            data.get('dividendYield'),
            data.get('per'),
            data.get('pbr'),
            data.get('marketCap'),
            data.get('volume'),
            movement,
            signal,
            ticker
        ))
        
        # 5. Insert into watch_detail
        # Use REPLACE INTO for MySQL (similar to INSERT OR REPLACE)
        cursor.execute('''
            REPLACE INTO watch_detail (
                ticker, yymmdd, companyName, currentPrice, previousClose,
                dividendYield, per, pbr, marketCap, volume,
                pricemovement, `signal_val`, memo
            ) VALUES (%s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s)
        ''', (
            ticker,
            today_str,
            data.get('companyName'),
            current_price,
            data.get('previousClose'),
            data.get('dividendYield'),
            data.get('per'),
            data.get('pbr'),
            data.get('marketCap'),
            data.get('volume'),
            movement,
            signal,
            "" # memo
        ))
        
    conn.commit()
    conn.close()
    print("Daily batch processing completed.")

if __name__ == '__main__':
    run_daily_batch()

File: internal/service/batch.go
package service

import (
	"context"
	"database/sql"
	"fmt"
	"log"
	"time"

	"smallcap-watcher-go/internal/api"
	"smallcap-watcher-go/internal/parse"
)

type WatchRow struct {
	Ticker string
}

func RunBatch(db *sql.DB, scraperBaseURL string) error {
	tickers, err := loadTickers(db)
	if err != nil {
		return err
	}

	client := api.NewClientWithBaseURL(scraperBaseURL)
	ctx := context.Background()

	for _, ticker := range tickers {
		payload, err := client.FetchStockData(ctx, ticker)
		if err != nil {
			log.Printf("failed to fetch %s: %v", ticker, err)
			continue
		}

		if err := updateStock(db, payload); err != nil {
			log.Printf("failed to update %s: %v", ticker, err)
		}
	}

	return nil
}

func loadTickers(db *sql.DB) ([]string, error) {
	rows, err := db.Query("SELECT ticker FROM watch_list")
	if err != nil {
		return nil, err
	}
	defer rows.Close()

	var tickers []string
	for rows.Next() {
		var ticker string
		if err := rows.Scan(&ticker); err != nil {
			return nil, err
		}
		tickers = append(tickers, ticker)
	}
	return tickers, rows.Err()
}

func updateStock(db *sql.DB, payload api.StockResponse) error {
	currentPriceVal, currentOK := parse.ParseNumeric(payload.CurrentPrice)
	pbrVal, pbrOK := parse.ParseNumeric(payload.PBR)
	volumeVal, volumeOK := parse.ParseNumeric(payload.Volume)
	previousCloseVal, prevOK := parse.ParsePreviousClose(payload.PreviousClose)

	var pricemovement string
	var signal string
	if currentOK && prevOK && previousCloseVal != 0 {
		movement := ((currentPriceVal - previousCloseVal) / previousCloseVal) * 100
		pricemovement = fmt.Sprintf("%.2f%%", movement)
		signal = generateSignal(movement)
	} else {
		pricemovement = ""
		signal = "Neutral"
	}

	now := time.Now().Format("060102")

	_, err := db.Exec(
		`INSERT INTO watch_list (
			ticker, companyName, currentPrice, previousClose, dividendYield, per, pbr, marketCap, volume, pricemovement, signal_val
		) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)
		ON DUPLICATE KEY UPDATE
			companyName=VALUES(companyName),
			currentPrice=VALUES(currentPrice),
			previousClose=VALUES(previousClose),
			dividendYield=VALUES(dividendYield),
			per=VALUES(per),
			pbr=VALUES(pbr),
			marketCap=VALUES(marketCap),
			volume=VALUES(volume),
			pricemovement=VALUES(pricemovement),
			signal_val=VALUES(signal_val)`,
		payload.Ticker,
		payload.CompanyName,
		nullFloat(currentPriceVal, currentOK),
		payload.PreviousClose,
		payload.DividendYield,
		payload.PER,
		nullFloat(pbrVal, pbrOK),
		payload.MarketCap,
		nullInt(volumeVal, volumeOK),
		pricemovement,
		signal,
	)
	if err != nil {
		return err
	}

	_, err = db.Exec(
		`REPLACE INTO watch_detail (
			ticker, yymmdd, companyName, currentPrice, previousClose, dividendYield, per, pbr, marketCap, volume, pricemovement, signal_val
		) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)`,
		payload.Ticker,
		now,
		payload.CompanyName,
		nullFloat(currentPriceVal, currentOK),
		payload.PreviousClose,
		payload.DividendYield,
		payload.PER,
		nullFloat(pbrVal, pbrOK),
		payload.MarketCap,
		nullInt(volumeVal, volumeOK),
		pricemovement,
		signal,
	)
	return err
}

func generateSignal(movement float64) string {
	if movement > 3.0 {
		return "Buy"
	}
	if movement < -3.0 {
		return "Sell"
	}
	return "Neutral"
}

func nullFloat(value float64, ok bool) sql.NullFloat64 {
	return sql.NullFloat64{Float64: value, Valid: ok}
}

func nullInt(value float64, ok bool) sql.NullInt64 {
	return sql.NullInt64{Int64: int64(value), Valid: ok}
}
